[PB] Fix WAIT_FOR_UPDATE_COMPLETE ioctl

NiLuJe opened the pull request:
Hopefully, fix #6000

Depends on https://github.com/koreader/koreader-base/pull/1188

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/6669)
<!-- Reviewable:end -->

NiLuJe commented:
Ideally, someone with a PB should double-check that I didn't horribly break all the things ;p.

(And that the logic actually works ^^).

NiLuJe commented:
That technically means that there's a good chance WAIT_FOR_COMPLETE has essentially always been returning -1 at best (or horribly crashing like in #6000) since basically forever on PB, I guess?

ezdiy commented:
> (And that the logic actually works ^^).

Any easy way to see that? Some showcase to mix up DU/A2/GL update in quick succession, see if EPDC fails to properly fence?

NiLuJe commented:
Just check debug logging with both this and https://github.com/koreader/koreader-base/pull/1188 applied, and check the early debug logging about the ioctl probing ;).

NiLuJe commented:
It'll be right after the fb setup recap, i.e.,

```
ffi.load: blitbuffer
FB: computed size: 3342336
FB: id           : mxc_epdc_fb
FB: smem_len     : 6684672
FB: line_length  : 1088
FB: xres         : 1080
FB: xres_virtual : 1088
FB: yres         : 1440
FB: yres_virtual : 3072
FB: bpp          : 8
09/17/20-19:35:41 INFO  mech_wait_update_complete: function: 0x2ad7d808
09/17/20-19:35:41 DEBUG kobo_mxc_wait_for_update_complete 1
09/17/20-19:35:41 INFO  mech_wait_update_complete(1): 0
```

Except, well, the PB version ;p.

NiLuJe commented:
(Hopefully the tests done in #6000 ensure that even if probing fails, it shouldn't crash).

ezdiy commented:
I've tried inverting the checks, ie both your patches as is, as well with:

```
-        self.mech_wait_update_complete = pocketbook_pb631_mxc_wait_for_update_complete
-        self.mech_wait_update_complete_fallback = pocketbook_mxc_wait_for_update_complete
+        self.mech_wait_update_complete_fallback = pocketbook_pb631_mxc_wait_for_update_complete
+        self.mech_wait_update_complete = pocketbook_mxc_wait_for_update_complete
```

To force its hand. In either case the check never triggers - meaning both ioctls return 0. Somewhat worrying whether we're selecting the right one (this is a new kernel, 3.10.65. The issues in #6000 are related to ancient 2.6/3.0 kernels IMO).

ezdiy commented:
Some crude strace on my device of vanilla inkview app:

```
ioctl(4, 0x4004462f, 0xa3015dc4)        = 0
ioctl(4, 0x80044655, 0xaedbd594)        = 0
```

So definitely it's the 0x2f (didn't manage to compile your strace due to some weird kernel header dep mishaps i was too lazy to troubleshoot, you don't happen to have static binary by any chance?)

NiLuJe commented:
@ezdiy: Huh. That's... strange, but doesn't surprise me all that much, the header and the driver code are already in disagreement in the PB631 sources... So it stands to reason that production kernels would be even more finicky...

But at least the ioctl should *definitely* return -EINVAL on an invalid command, so if both work, and they at least didn't break the API, I'll take that as a win?

FWIW, the test works if I force an unsupported ioctl on Kobo:

```
09/17/20-21:31:30 INFO  mech_wait_update_complete: function: 0x2ad7c690
09/17/20-21:31:30 DEBUG kobo_mk7_mxc_wait_for_update_complete 1
09/17/20-21:31:30 INFO  mech_wait_update_complete(1): -1
```

----

As far as strace is concerned, I never did build one for PB (I don't even have a PB-specific patch ^^), so I don't have one for you, sorry :/. But, yeah, it's a PITA to build.

NiLuJe commented:
TL;DR: I guess we'd need someone affected by #6000 to test this?

NiLuJe commented:
A quick test you could do with an unpatched strace is to enable the timing flag (-T), and check that it actually blocks for ~500ms:

```
strace: Process 9554 attached
strace: Process 9553 attached
[pid  9553] 21:39:32.226016 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=831}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_FULL, update_marker=4, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x2bc169c8) = 0 <0.001739>
[pid  9553] 21:39:32.230651 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {4}, 0x2ad92f30) = 0x3b2 <0.545623>
[pid  9553] 21:39:32.810884 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=27, left=402, width=130, height=72}, waveform_mode=NTX_WFM_MODE_A2, update_mode=UPDATE_MODE_PARTIAL, update_marker=5, temp=TEMP_USE_AMBIENT, flags=EPDC_FLAG_FORCE_MONOCHROME, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x2c0e2d38) = 0 <0.001946>
[pid  9553] 21:39:32.887892 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=1009}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_PARTIAL, update_marker=6, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x3082b770) = 0 <0.002920>
[pid  9553] 21:39:34.185196 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {6}, 0x2d2fada0) = 0 <0.000044>
[pid  9553] 21:39:34.185935 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=1429}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_FULL, update_marker=7, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x30896110) = 0 <0.001622>
[pid  9553] 21:39:34.188974 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {7}, 0x2c080b48) = 0x3b5 <0.510740>
[pid  9553] 21:39:37.266455 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=1009}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_FULL, update_marker=8, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x3092d248) = 0 <0.001740>
[pid  9553] 21:39:37.270755 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {8}, 0x2c0d7ce8) = 0x3b2 <0.546752>
[pid  9553] 21:39:38.159506 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=1429}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_FULL, update_marker=9, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x30887030) = 0 <0.001640>
[pid  9553] 21:39:38.162687 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {9}, 0x30a1fb48) = 0x3b4 <0.517105>
[pid  9553] 21:39:40.202865 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=1009}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_FULL, update_marker=10, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x309cd328) = 0 <0.001737>
[pid  9553] 21:39:40.206325 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {10}, 0x2b13e648) = 0x3b2 <0.541082>
[pid  9553] 21:39:40.983140 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=1429}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_FULL, update_marker=11, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x308b6d30) = 0 <0.001614>
[pid  9553] 21:39:40.987396 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {11}, 0x2c048ad0) = 0x3b4 <0.512402>
[pid  9553] 21:39:43.647627 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=1009}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_FULL, update_marker=12, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x309218d0) = 0 <0.002468>
[pid  9553] 21:39:43.651604 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {12}, 0x2d2df8a0) = 0x3b2 <0.545716>
[pid  9553] 21:39:44.337240 [2ac7a4ec] ioctl(4, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=11, left=0, width=1080, height=1429}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_FULL, update_marker=13, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x309a0e80) = 0 <0.002282>
[pid  9553] 21:39:44.341056 [2ac7a4ec] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {13}, 0x2c0fcdd0) = 0x3b4 <0.518713>
```

NiLuJe commented:
Might be a safer bet to just assume `_IOWR('F', 0x2F, struct mxcfb_update_marker_data)` everywhere?

ezdiy commented:
The update ioctls are lgtm (at worst, it wont make things worse than it were and it's a safety for kernels where the ioctl would fail). The other stuff that's mixed in is a bit suspect tho.

Since we're getting bold here, some other tweaks:

```
--- a/ffi/framebuffer_mxcfb.lua
+++ b/ffi/framebuffer_mxcfb.lua
@@ -485,9 +485,9 @@ local function refresh_pocketbook(fb, refreshtype, waveform_mode, x, y, w, h)
     -- Enable the appropriate flag when requesting a REAGLD waveform (EPDC_WFTYPE_AAD on PB631)
     if waveform_mode == C.EPDC_WFTYPE_AAD then
         refarea[0].flags = C.EPDC_FLAG_USE_AAD
-    elseif waveform_mode == C.WAVEFORM_MODE_A2 then
-        --- @fixme: fast is currently set to DU, not A2!
-        -- As well as when requesting a 2bit waveform
+    elseif waveform_mode == C.WAVEFORM_MODE_DU then
+        --- @fixme: lower latency, turns out its the same as remarkable
+        refarea[0].temp = 0x18
         refarea[0].flags = C.EPDC_FLAG_FORCE_MONOCHROME
     else
```

Also, the following essentially invoke something reagl-like in the driver:

```
        self.waveform_fast = C.WAVEFORM_MODE_DU
        self.waveform_ui = C.WAVEFORM_MODE_GL16
        self.waveform_flashui = C.WAVEFORM_MODE_GC16
        self.waveform_full = C.WAVEFORM_MODE_GC16
        self.waveform_partial = C.WAVEFORM_MODE_GL16
        self.waveform_night = C.WAVEFORM_MODE_GC16
        self.waveform_flashnight = self.waveform_night
        self.night_is_reagl = false
```

And

```
--- a/frontend/device/pocketbook/device.lua
+++ b/frontend/device/pocketbook/device.lua
@@ -36,9 +36,6 @@ local PocketBook = Generic:new{
 
     -- all devices that have warmth lights use inkview api
     hasNaturalLightApi = yes,
-
-    -- Let's assume that HW inversion is generally safe
-    canHWInvert = yes,
 }
 
 -- Make sure the C BB cannot be used on devices with a 24bpp fb
@@ -116,7 +113,7 @@ function PocketBook:init()
                 end
             elseif ev.code == C.EVT_FOREGROUND or ev.code == C.EVT_SHOW then
                 tryOpenBook()
-                ui:setDirty('all', 'partial')
+                ui:setDirty('all', 'fast')
                 if quasiSuspended then
                     quasiSuspended = false
                     return "Resume"
```

HWInvert is wishful thinking, only the soft one works. The "fast" one is a hack, as we're guaranteed to have a white foreground in there, so it always renders ok w/o flash. With all of the above, flashes are eliminated altogether from bogus places (unless explicitly asked for via full page refresh or local UI button click).

ezdiy commented:
> Might be a safer bet to just assume `_IOWR('F', 0x2F, struct mxcfb_update_marker_data)` everywhere?

If y you mean sony, carvantes etc ... well if spurious crashes happen there might be worth it. I presume the ifdef MX50 crap + ioctl nums + struct layout is same for all of those along with pocketbook.

NiLuJe commented:
> > Might be a safer bet to just assume `_IOWR('F', 0x2F, struct mxcfb_update_marker_data)` everywhere?
> 
> If y you mean sony, carvantes etc ... well if spurious crashes happen there might be worth it. I presume the ifdef MX50 crap + ioctl nums + struct layout is same for all of those along with pocketbook.

No, I meant on PB ;).

To that effect, I wouldn't mind you checking [this](https://github.com/koreader/koreader/pull/6669#issuecomment-694458006) to see if the "new" default actually doe anything useful ^^

ezdiy commented:
Fencing in native inkview reader:

```
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000080>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000075>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.186024>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.053210>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000089>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000087>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000073>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000085>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000076>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000081>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000070>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000082>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000080>
[a1645d7c] ioctl(4, 0x4004462f, 0xa3015dc4) = 0 <0.000082>
```

Rapid scrolling, one can see the occasional 50-150ms blib.

Koreader:

```
[3b205d7c] ioctl(4, 0xc008462f, 0x2e58078) = 0 <0.240199>
[3b205d7c] ioctl(4, 0xc008462f, 0x30489c0) = 0 <0.000081>
[3b205d7c] ioctl(4, 0xc008462f, 0x306ffb8) = 0 <0.000094>
[3b205d7c] ioctl(4, 0xc008462f, 0x32dc788) = 0 <0.000134>
[3b205d7c] ioctl(4, 0xc008462f, 0x308f478) = 0 <0.330867>
```

However:

```
[3b1f5d7c] ioctl(4, 0xc0084635, 0x1fb9460) = 0 <0.000107>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x1281dd8) = 0 <0.000319>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x119e448) = 0 <0.000099>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x11efe38) = 0 <0.000096>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x1d503e0) = 0 <0.000085>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x21280d8) = 0 <0.000124>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x1dfc830) = 0 <0.000188>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x207b668) = 0 <0.000108>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x207ed68) = 0 <0.000113>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x1fa1190) = 0 <0.000101>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x20d8dd8) = 0 <0.000148>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x20bac30) = 0 <0.000102>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x20ad728) = 0 <0.000132>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x20d9d90) = 0 <0.000105>
[3b1f5d7c] ioctl(4, 0xc0084635, 0xfdc858) = 0 <0.000105>
[3b1f5d7c] ioctl(4, 0xc0084635, 0x20c85d0) = 0 <0.000133>
```

Ie the alt 0x35 mode seems like a NOP always.

ezdiy commented on frontend/device/pocketbook/device.lua:
```diff
@@ -37,13 +37,14 @@ local PocketBook = Generic:new{
 -- Make sure the C BB cannot be used on devices with a 24bpp fb
 function PocketBook:blacklistCBB()
     -- As well as on those than can't do HW inversion, as otherwise NightMode would be ineffective.
+    --- @note: Since HWInvert is a no-go on PB, the C BB is essentially *always* blacklisted.
```
It might be worth investigating flipping the BB on the fly depending on nightmode setting.

NiLuJe replied:
That's... mildly annoying, IIRC, there's no runtime switch possible currently, it's handled at require time.

NiLuJe commented:
@ezdiy: Huh, thanks!

I'll try to prepare a strace patch + build to put our minds at ease (and make this kind of stuff easier to RE in the future).

NiLuJe commented:
Okay, here goes a strace build that should hopefully decode those ioctls without imploding ;).

[strace-pocketbook.tar.gz](https://github.com/koreader/koreader/files/5242439/strace-pocketbook.tar.gz)

NiLuJe commented:
FWIW, with default settings, closing the top menu should trigger a wait_for (basically, anything that does a full-screen flash should).

ezdiy commented:
So theory what really goes on. The header

```
#ifdef MX50_IOCTL_IF//[
#define MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631    _IOW('F', 0x2F, __u32)
#define MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631_V2 _IOWR('F', 0x35, struct mxcfb_update_marker_data)
#else //][!MX50_IOCTL_IF
#define MXCFB_WAIT_FOR_UPDATE_COMPLETE_ANDROID _IOWR('F', 0x35, struct mxcfb_update_marker_data)
#endif//] MX50_IOCTL_IF
```

is probably incorrect. My kernel is indeed `_IOW('F', 0x2F, __u32)` confirmed by your strace. On older kernels, the value is most likely `_IOWR('F', 0x2F, struct mxcfb_update_marker_data)` however. What then happens is the driver ignoring the struct size portion or just writing back whatever. Alternatively, the ioctl is always just `_IOW('F', 0x2F, __u32)`, and older kernels writing back the struct (ie kernel doesn't really adhere to ioctl convention anymore). Utterly broken, but not really surprising. This should get confirmed if somobody does strace in #6000 

Also, some long-wanted info about how waveform is set up:
```
[pid  1015] 10:39:00.592865 [  54] [a1645d7c] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631, {1409287144}, 0xa3015dc4) = 0 <0.000126>
[pid  1015] 10:39:00.593702 [  54] [a1645d7c] ioctl(4, MXCFB_SEND_UPDATE, {update_region={top=0, left=0, width=1872, height=1404}, waveform_mode=EPDC_WFTYPE_GC16, update_mode=UPDATE_MODE_FULL, update_marker=1426064360, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x9e589d60) = 0 <0.033861>
[pid  1000] 10:39:01.827624 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x55, 0x4), 0xaedbd784) = 0 <0.001620>
[pid  1000] 10:39:03.375642 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x55, 0x4), 0xaedbd784) = 0 <0.000512>
[pid  1000] 10:39:04.917135 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x55, 0x4), 0xaedbd784) = 0 <0.001653>
[pid  1000] 10:39:06.216783 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x56, 0x20), 0xaedbac48) = 0 <0.000518>
[pid  1000] 10:39:06.536894 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x55, 0x4), 0xaedbd594) = 0 <0.000108>
[pid  1000] 10:39:06.609980 [  54] [a1645d7c] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631, {1426064360}, 0xa3015dc4) = 0 <0.000723>
[pid  1015] 10:39:06.618106 [  54] [a1645d7c] ioctl(4, FBIOGET_VSCREENINFO, 0x9e589c14) = 0 <0.000165>
[pid  1015] 10:39:06.618696 [  54] [a1645d7c] ioctl(4, FBIOGET_VSCREENINFO, 0x9e589cb4) = 0 <0.000154>
[pid  1015] 10:39:06.619269 [  54] [a1645d7c] ioctl(4, FBIOGET_FSCREENINFO, 0x9e589d54) = 0 <0.000131>
[pid  1015] 10:39:06.653681 [  54] [a1645d7c] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631, {1426064360}, 0xa3015dc4) = 0 <0.000122>
[pid  1015] 10:39:06.654504 [  54] [a1645d7c] ioctl(4, MXCFB_SEND_UPDATE, {update_region={top=0, left=0, width=1872, height=1404}, waveform_mode=EPDC_WFTYPE_GC16, update_mode=UPDATE_MODE_PARTIAL, update_marker=1442841576, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x9e589d60) = 0 <0.024451>
```

Essentially, only EPDC_WFTYPE_GC16{mode=PARTIAL} vs EPDC_WFTYPE_GC16{mode=FULL} is used, where the partial one doesn't flash. No other ioctls are observed here (!).

NiLuJe commented:
Okay, that's indeed what I was starting to suspect, thanks ;).

But busy this afternoon, will clean this up later this evening.

On Fri, Sep 18, 2020, 15:19 ezdiy <notifications@github.com> wrote:

> So theory what really goes on. The header
>
> #ifdef MX50_IOCTL_IF//[
> #define MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631    _IOW('F', 0x2F, __u32)
> #define MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631_V2 _IOWR('F', 0x35, struct mxcfb_update_marker_data)
> #else //][!MX50_IOCTL_IF
> #define MXCFB_WAIT_FOR_UPDATE_COMPLETE_ANDROID _IOWR('F', 0x35, struct mxcfb_update_marker_data)
> #endif//] MX50_IOCTL_IF
>
> is probably incorrect. My kernel is indeed _IOW('F', 0x2F, __u32)
> confirmed by your strace. On older kernels, the value is most likely _IOWR('F',
> 0x2F, struct mxcfb_update_marker_data) however. What then happens is the
> driver ignoring the struct size portion or just writing back whatever.
> Alternatively, the ioctl is always just _IOW('F', 0x2F, __u32), and older
> kernels writing back the struct (ie kernel doesn't really adhere to ioctl
> convention anymore). Utterly broken, but not really surprising. This should
> get confirmed if somobody does strace in #6000
> <https://github.com/koreader/koreader/issues/6000>
>
> Also, some long-wanted info about how waveform is set up:
>
> [pid  1015] 10:39:00.592865 [  54] [a1645d7c] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631, {1409287144}, 0xa3015dc4) = 0 <0.000126>
> [pid  1015] 10:39:00.593702 [  54] [a1645d7c] ioctl(4, MXCFB_SEND_UPDATE, {update_region={top=0, left=0, width=1872, height=1404}, waveform_mode=EPDC_WFTYPE_GC16, update_mode=UPDATE_MODE_FULL, update_marker=1426064360, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x9e589d60) = 0 <0.033861>
> [pid  1000] 10:39:01.827624 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x55, 0x4), 0xaedbd784) = 0 <0.001620>
> [pid  1000] 10:39:03.375642 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x55, 0x4), 0xaedbd784) = 0 <0.000512>
> [pid  1000] 10:39:04.917135 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x55, 0x4), 0xaedbd784) = 0 <0.001653>
> [pid  1000] 10:39:06.216783 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x56, 0x20), 0xaedbac48) = 0 <0.000518>
> [pid  1000] 10:39:06.536894 [  54] [a1645d7c] ioctl(4, _IOC(_IOC_READ, 0x46, 0x55, 0x4), 0xaedbd594) = 0 <0.000108>
> [pid  1000] 10:39:06.609980 [  54] [a1645d7c] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631, {1426064360}, 0xa3015dc4) = 0 <0.000723>
> [pid  1015] 10:39:06.618106 [  54] [a1645d7c] ioctl(4, FBIOGET_VSCREENINFO, 0x9e589c14) = 0 <0.000165>
> [pid  1015] 10:39:06.618696 [  54] [a1645d7c] ioctl(4, FBIOGET_VSCREENINFO, 0x9e589cb4) = 0 <0.000154>
> [pid  1015] 10:39:06.619269 [  54] [a1645d7c] ioctl(4, FBIOGET_FSCREENINFO, 0x9e589d54) = 0 <0.000131>
> [pid  1015] 10:39:06.653681 [  54] [a1645d7c] ioctl(4, MXCFB_WAIT_FOR_UPDATE_COMPLETE_PB631, {1426064360}, 0xa3015dc4) = 0 <0.000122>
> [pid  1015] 10:39:06.654504 [  54] [a1645d7c] ioctl(4, MXCFB_SEND_UPDATE, {update_region={top=0, left=0, width=1872, height=1404}, waveform_mode=EPDC_WFTYPE_GC16, update_mode=UPDATE_MODE_PARTIAL, update_marker=1442841576, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x9e589d60) = 0 <0.024451>
>
> Essentially, only EPDC_WFTYPE_GC16{mode=PARTIAL} vs
> EPDC_WFTYPE_GC16{mode=FULL} is used, where the partial one doesn't flash.
> No other ioctls are observed here (!).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/koreader/koreader/pull/6669#issuecomment-694863258>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3KZRAL36I4AZXAT5R5NLSGNM7JANCNFSM4RQZDD7Q>
> .
>

ezdiy commented:
@NiLuJe So it's the "utterly broken" variant. I'd revert all the detection logic (u32 ioctl, yet kernel expecting writable struct sometimes). It should suffice to just stick with what https://github.com/koreader/koreader-base/pull/1187 did. And just keep single

```
// CAVEAT: Some kernels expect struct mxcfb_update_marker_data regardless in here, so ioctl should always pass that.
#define MXCFB_WAIT_FOR_UPDATE_COMPLETE    _IOW('F', 0x2F, __u32)
```

in the cdecl header.

NiLuJe commented:
@ezdiy: Yep! WIll do! (FFTH setup day, things were a bit hectic today ^^).

ezdiy commented:
[ACK](https://github.com/koreader/koreader-base/pull/1188) (running it for a day now).

NiLuJe commented:
I just realized when seeing the FBInk output, that, in all the PB strace logs we've seen, even when it does appear to block, it's doing so in a highly irregular way: it's only ever supposed to return 0 when it *didn't have to* block (i.e., the ioctl returns the amount of jiffies the call blocked for).

That's... weird. Because that's a handled via a standard `wait_for_completion_timeout` kernel API... :?

Oh, well :D.

Oh, well :D.

NiLuJe commented:
By "the FBInk output", I meant that in verbose mode, FBInk handles that return value to make it human-readable, so, even the tiny refresh done by the test command I asked for (`fbink -vMmw "Hello"`) should cause *some* amount of blocking.

i.e., on a non-REAGL Kobo:

```
Waited 320ms for completion of update 11338
```

Instead of 

```
Update 15656 has already fully been completed
```

:s

ezdiy commented:
My device is indeed fully raegal no matter what I do, as in, no blocking is observed in practice - the display is slow, sure, but the thread rolls forward and only when several refreshes are in the blocking will start showing up. Inkview *does* the wait in a separate thread though, so older devices are non-raegal for sure (and perhaps where the ioctl will ultimately matter).

NiLuJe commented:
FWIW, I've always seen this block as expected (i.e., if you wait for a long-ago marker, it doesn't and return 0) and return > 0 on everything I could get my hands on (so, mainly Kindle & Kobos), REAGL or not, which is whyI find the return value shenanigan curious ^^.

But, if the actual call does block, as the strace -T log seem to suggest, eh, that'll do ;).

NiLuJe commented:
e.g., if I run FBInk itself under strace, it matches:

```
┌─(ROOT@dahlia:pts/1)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(0.20:21%:16:28:63%:#)── strace -fitvT -e trace=ioctl fbink -vMmw "Hello"                                                                                                                                                                                                                                                                            ──(Sat, Sep 19)─┘
[FBInk] Detected a Kobo Aura H2O (370 => Dahlia @ Mark 5)
[FBInk] Clock tick frequency appears to be 100 Hz
[FBInk] Screen density set to 265 dpi
16:46:16 [2abbc4ec] ioctl(3, FBIOGET_VSCREENINFO, 0x107390) = 0 <0.000046>
[FBInk] Variable fb info: 1080x1440, 8bpp @ rotation: 1 (Clockwise, 90°)
[FBInk] Enabled Kobo viewport insanity (1080x1440 -> 1080x1429), top-left corner is @ (0, 11)
[FBInk] Fontsize set to 24x24 (IBM base glyph size: 8x8)
[FBInk] Line length: 45 cols, Page size: 59 rows
[FBInk] Horizontal fit is perfect!
[FBInk] Vertical fit isn't perfect, shifting rows down by 7 pixels
16:46:16 [2abbc4ec] ioctl(3, FBIOGET_FSCREENINFO, 0x107430) = 0 <0.000028>
[FBInk] Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 6684672 bytes & line length: 1088 bytes
[FBInk] Pen colors set to #000000 for the foreground and #FFFFFF for the background
Printing string 'Hello' @ column 0 + 0px, row 0 + 0px (overlay: N, no BG: N, no FG: N, inverted: N, flashing: N, centered: Y, halfway: Y, left padded: N, right padded: N, clear screen: N, waveform: AUTO, dithering: PASSTHROUGH, nightmode: N, skip refresh: N, font: 0, font scaling: x0)
Adjusted row to 29 for vertical centering
Adjusted position: column 0, row 29
Final position: column 0, row 29
Need 1 lines to print 5 characters over 43 available columns
Line 1 (of ~1), previous line was 0 characters long and there were 5 characters left to print
Characters to print: 5 out of the 5 remaining ones
Line takes up 5 bytes
Adjusted column to 20 for centering
snprintf wrote 5 bytes
Printing 'Hello' @ line offset 0 (meaning row 29)
Character count: 5 (over 5 bytes)
Adjusting vertical pen position by 0 pixels, as requested, plus 18 pixels, as mandated by the native viewport
Region: top=714, left=480, width=120, height=24
Char 1 out of 5 is @ byte offset 0 and is U+0048 (H)
Char 2 out of 5 is @ byte offset 1 and is U+0065 (e)
Char 3 out of 5 is @ byte offset 2 and is U+006C (l)
Char 4 out of 5 is @ byte offset 3 and is U+006C (l)
Char 5 out of 5 is @ byte offset 4 and is U+006F (o)
16:46:16 [2abbc4ec] ioctl(3, MXCFB_SEND_UPDATE_V1_NTX, {update_region={top=714, left=480, width=120, height=24}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_PARTIAL, update_marker=11361, temp=TEMP_USE_AMBIENT, flags=0, alt_buffer_data={virt_addr=(nil), phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7ed47a90) = 0 <0.004205>
waveform_mode is now 0x101 (AUTO)
Next line should start @ row 1
16:46:16 [2abbc4ec] ioctl(3, MXCFB_WAIT_FOR_UPDATE_COMPLETE_V1, {11361}, 0x7ed47ba8) = 0x3cb <0.289732>
Waited 290ms for completion of update 11361
16:46:17 [????????] +++ exited with 0 +++
```

`0x3cb` is the block time in jiffies, and, knowing the timeout value on Kobo kernels and the crap HZ resolution, that rounds to about 290ms, which agrees with strace's actual measurement at 289.7mls ;).

EastEriq commented:
I've OTA-pulled the latest development build on my PB631, and confirm that koreader rocks. No crash nor missing functionality observed while fiddling around for quite some time.
